crates/api/src/error.rs

```diff
@@ -0,0 +1,127 @@
+//! Kitsune2 error types.
+
+use std::sync::Arc;
+
+/// A clonable trait-object inner error.
+#[derive(Clone, Default)]
+pub struct DynInnerError(
+    pub Option<Arc<dyn std::error::Error + 'static + Send + Sync>>,
+);
+
+impl std::fmt::Debug for DynInnerError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl std::fmt::Display for DynInnerError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.0.as_ref() {
+            None => f.write_str("None"),
+            Some(s) => s.fmt(f),
+        }
+    }
+}
+
+impl std::error::Error for DynInnerError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        self.0.as_ref().map(|s| {
+            let out: &(dyn std::error::Error + 'static) = &**s;
+            out
+        })
+    }
+}
+
+impl DynInnerError {
+    /// Construct a new DynInnerError from a source error.
+    pub fn new<E: std::error::Error + 'static + Send + Sync>(e: E) -> Self {
+        Self(Some(Arc::new(e)))
+    }
+}
+
+/// The core kitsune2 error type. This type is used in all external
+/// kitsune apis as well as internally in some modules.
+///
+/// This type is required to implement `Clone` to ease the use of
+/// shared futures, which require the entire `Result` to be `Clone`.
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum K2Error {
+    /// Generic Kitsune2 internal error.
+    #[error("{ctx} (src: {src})")]
+    Other {
+        /// Any context associated with this error.
+        ctx: Arc<str>,
+
+        /// The inner error (if any).
+        #[source]
+        src: DynInnerError,
```

Comment:
Will the inner type ever be anything other than a `std::io::Error`? If not then can we just make this `Option<std::io::Error>` or if there's a small chance that we might add another error type later then maybe we could call this variant `OtherIo` and still make the change?

Reply:
I think there's a good chance we want other error types. I'd put some error handling into my gossip code just so I could keep moving, not necessarily planning to keep it. I was using a similar trait definition to this one and it's been useful so far :)

Reply:
Yeah, this other type is the general fallback case. Everything from io errors to every kind of dependency library error, to our own internal custom errors can go in here until we identify the need to break one/any of those out into their own top-level variants.

Reply:
Huh interesting point. Yeah that does make sense